### PR TITLE
Different dimension for attention

### PR DESCRIPTION
**Pull request by pggPL**
# Description

I added `channels_per_head` parameter to attention. I know that it is breaking change, but I think there is not a good way of doing it - look at the discussion in https://github.com/NVIDIA/TransformerEngine/pull/731

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

Please list the changes introduced in this PR:

- Added `channels_per_head` parameter to attention,
- Added simple unit test, I don't know if its necessary

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb
Comment (ptrendx) [inline docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb:1]: I think this is unintended change?

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -3207,6 +3207,7 @@ def __init__(
         cp_group: Optional[dist_group_type] = None,
         cp_global_ranks: List[int] = None,
         cp_stream: torch.cuda.Stream = None,
+        kv_channels: int = None # deprecated
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:3210]: Well, if it is deprecated, then we should make sure that the code that explicitly calls `kv_channels=something` still works as before, right? While those changes are just ignoring it.

File: docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb
Comment (pggPL) [reply docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb:1]: Yes, I will remove it.

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (pggPL) [reply transformer_engine/pytorch/attention.py:3210]: I think the code will crash when kv_channels were set for different value than default, so we can safely do this. We have in the code 
```
self.hidden_size_kv = int(hidden_size * self.num_gqa_groups // num_attention_heads)
...
self.layernorm_qkv = LayerNormLinear(
                    hidden_size,
                    hidden_size + 2 * self.hidden_size_kv,
                    ...
```

So the input size to the attention does not depend from kv_channels - do kv_channels can have one value.

I see this is not the case with DotProductAttention, I will look at this. 

Nevertheless, I want to ask if I want to deprecate an argument, what should I do? Will causing it to behave the same at previously, but removing from docs will be ok? Do we consider it as breaking the API?

File: docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb
Comment (pggPL) [reply docs/examples/te_llama/tutorial_accelerate_hf_llama_with_te.ipynb:1]: No, apparently it's ok. In last commit which was merged it was accidentally changed, so we want to revert it.

File: tests/pytorch/test_gqa.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        ffn_hidden_size,
+        num_attn_head,
+        num_gqa_groups,
+        kv_channels=kv_channels
+    )
+
+    # Run forward pass
+    x = torch.randn((batch_size, 1, hidden_size)).cuda()
+    model(x)
+
```
Comment (sudhakarsingh27) [inline tests/pytorch/test_gqa.py:38]: Maybe also add a check for the dimensions (`hidden_size_q` and `hidden_size_kv`) of the underlying weight tensors

**Review by sudhakarsingh27 — COMMENTED** [review]
The tutorial file's mode has been changed. Could you pls remove that change?

**Comment by pggPL** [comment]
But this tutorial mode change is good. Should I remove it and add to the new PR?

**Review by sudhakarsingh27 — APPROVED** [review]

**Review by ksivaman — COMMENTED** [review]
Why change the file permission for the tutorial?

**Comment by sudhakarsingh27** [comment]
> Why change the file permission for the tutorial?

It seems like we mistakenly changed the permission last time so I think we're reverting it back. Is that okay or do we need to create a separate fix for it?

**Pull request merged**